pkg/importer/imageio-datasource.go

```diff
@@ -0,0 +1,508 @@
+	diskResponse, err := diskRequest.Send()
+	if err != nil {
+		return nil, int64(0), errors.Wrap(err, "Error fetching disk")
+	}
+	disk, success := diskResponse.Disk()
+	if !success {
+		return nil, int64(0), errors.New("Error disk not found")
+	}
+
+	totalSize, available := disk.TotalSize()
+	if !available {
+		return nil, int64(0), errors.New("Error total disk size not available")
+	}
+
+	id, available := disk.Id()
+	if !available {
+		return nil, int64(0), errors.New("Error disk id not available")
+	}
+
+	image, err := ovirtsdk4.NewImageBuilder().Id(id).Build()
+	if err != nil {
+		return nil, int64(0), errors.Wrap(err, "Error building image object")
+	}
+
+	transfersService := conn.SystemService().ImageTransfersService()
+	transfer := transfersService.Add()
+	imageTransfer, err := ovirtsdk4.NewImageTransferBuilder().Image(
+		image,
+	).Direction(
+		ovirtsdk4.IMAGETRANSFERDIRECTION_DOWNLOAD,
+	).Build()
+	if err != nil {
+		return nil, int64(0), errors.Wrap(err, "Error preparing transfer object")
+	}
+
+	transfer.ImageTransfer(imageTransfer)
+	var it = &ovirtsdk4.ImageTransfer{}
+	for {
+		response, err := transfer.Send()
+		if err != nil {
+			if strings.Contains(err.Error(), "Disk is locked") {
+				klog.Infoln("waiting for disk to unlock")
+				time.Sleep(15 * time.Second)
+				continue
+			}
+			return nil, int64(0), errors.Wrap(err, "Error sending transfer image request")
+		}
+		it, available = response.ImageTransfer()
+		if !available {
+			return nil, int64(0), errors.New("Error image transfer not available")
+		}
+		phase, available := it.Phase()
+		if !available {
+			return nil, int64(0), errors.New("Error phase not available")
+		}
+		if phase == ovirtsdk4.IMAGETRANSFERPHASE_INITIALIZING {
+			time.Sleep(1 * time.Second)
+		} else if phase == ovirtsdk4.IMAGETRANSFERPHASE_TRANSFERRING {
+			break
+		} else {
+			return nil, int64(0), errors.Errorf("Error transfer phase: %s", phase)
+		}
+	}
+	return it, totalSize, nil
+}
+
+func loadCA(certDir string) (*x509.CertPool, error) {
+	if certDir == "" {
+		return nil, errors.New("Error CA not provided")
+	}
+	files, err := ioutil.ReadDir(certDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error listing files in %s", certDir)
+	}
+
+	caCertPool := x509.NewCertPool()
+	for _, file := range files {
+		if file.IsDir() || file.Name()[0] == '.' {
+			continue
+		}
+
+		fp := path.Join(certDir, file.Name())
+
+		klog.Infof("Attempting to get certs from %s", fp)
+
+		certs, err := ioutil.ReadFile(fp)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error reading file %s", fp)
+		}
+
+		if ok := caCertPool.AppendCertsFromPEM(certs); !ok {
+			klog.Warningf("No certs in %s", fp)
+		}
+	}
+	return caCertPool, nil
+}
+
+// may be overridden in tests
+var newOvirtClientFunc = getOvirtClient
+
+// Not well defined abstractions in the SDK so we need to define below interfaces to mock the calls
+
+// ConnectionInterface defines connection methods
+type ConnectionInterface interface {
```

Comment:
I am a little concerned these aren't defined in the SDK, and everyone has to define them. Is there no better way to do this?

Reply:
Unfortunately SDK code is generated and the interfaces are not. In order to be able to mock the code in test we had to implement them here. We can open an issue for this but I doubt it will be fixed anytime soon.

Reply:
Yes, that's not defined in SDK. No way that I am aware of. IMHO the best would be have those mocks inside the SDK itself.

Reply:
I created this issue https://github.com/oVirt/ovirt-engine-sdk-go/issues/199 to track this change.